### PR TITLE
[FLINK-25645][table-runtime] UnsupportedOperationException would thrown out when hash shuffle by a field with array type

### PR DESCRIPTION
**Pull request by lsyldliu**
## What is the purpose of the change
Support complex shuffle hash key type

## Brief change log

  - *Support complex shuffle hash key type*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added unit tests in HashCodeGeneratorTest*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: ( no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by lsyldliu** [comment]
@beyond1920 Can you help review this pr?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala [hunk 1]
```diff
@@ -344,6 +353,34 @@ object CodeGenUtils {
...
+    ctx.addReusableInitStatement(s"$hashFunc = new ${genHash.getClassName}($refs);")
+    s"$hashFunc.hashCode($term)"
+  }
+
+  def genHashForMap(
+      ctx: CodeGeneratorContext,
+      keyType: LogicalType,
+      valueType: LogicalType,
+      term: String): String = {
+    val subCtx = CodeGeneratorContext(ctx.tableConfig)
```
Comment (beyond1920) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala:373]: The code is duplicate with `genHashForArray`, could we avoid this?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/HashCodeGenerator.scala [hunk 2]
```diff
@@ -75,6 +75,146 @@ object HashCodeGenerator {
...
+
+        ${ctx.reuseMemberCode()}
+
+        public $className(Object[] references) throws Exception {
+          ${ctx.reuseInitCode()}
+        }
+
+        @Override
+        public int hashCode($ARRAY_DATA $inputTerm) {
+          int $hashIntTerm = 0;
```
Comment (beyond1920) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/HashCodeGenerator.scala:123]: Did you forget to call `ctx.reuseLocalVariableCode()` here?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/HashCodeGenerator.scala [hunk 3]
```diff
@@ -75,6 +75,146 @@ object HashCodeGenerator {
...
+
+        ${ctx.reuseMemberCode()}
+
+        public $className(Object[] references) throws Exception {
+          ${ctx.reuseInitCode()}
+        }
+
+        @Override
+        public int hashCode($MAP_DATA $inputTerm) {
+          $ARRAY_DATA $keys = $inputTerm.keyArray();
```
Comment (beyond1920) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/HashCodeGenerator.scala:187]: Did you forget to call `ctx.reuseLocalVariableCode()` here?

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/codegen/HashCodeGeneratorTest.scala [hunk 4]
```diff
@@ -18,18 +18,26 @@
...
+import org.junit.rules.ExpectedException
+
+import scala.collection.JavaConversions.mapAsJavaMap
 
 /** Test for [[HashCodeGenerator]]. */
 class HashCodeGeneratorTest {
 
   private val classLoader = Thread.currentThread().getContextClassLoader
 
+  var expectedException: ExpectedException = ExpectedException.none
```
Comment (beyond1920) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/codegen/HashCodeGeneratorTest.scala:34]: Would you please add a ITCase in SQL to cover group by array type fields and map type fields?

**Review by beyond1920 — COMMENTED** [review]
@lsyldliu Thanks for contribution.
LGTM overall, I only left some minor comments.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/HashCodeGenerator.scala (hunk 2 above)
Comment (lsyldliu) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/HashCodeGenerator.scala:123]: Here all the local variables are declared explicitly, so it is no need to call ctx.reuseLocalVariableCode() .

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/codegen/HashCodeGeneratorTest.scala (hunk 4 above)
Comment (lsyldliu) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/codegen/HashCodeGeneratorTest.scala:34]: I have add an ITCase about array type in group by clause.

Comment (lsyldliu) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/codegen/HashCodeGeneratorTest.scala:34]: I have researched spark and presto. In spark, map type is not allowed used in join and group by clause because of map type is not an orderable type, but it supported as the distribute by key. In presto, map supported as group by key, it is because presto only implement hash-based aggregate. So, I think we should support map type field as shuffle key, but doesn't as sort key, such as order by/sort-base join/agg clause. we should support to generate hash code for map which is also as spark and presto, this may be useful for distribute by clause which we will support in 1.16.

**Comment by lsyldliu** [comment]
@beyond1920 Thanks for your review, I have updated the pr according to your feedback.

**Comment by lsyldliu** [comment]
@flinkbot run azure

**Comment by lsyldliu** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/HashCodeGenerator.scala [hunk 5]
```diff
@@ -75,6 +75,157 @@ object HashCodeGenerator {
...
+            boolean $keyIsNull = $keys.isNullAt($i);
+            if (!$keyIsNull) {
+              $keyHashTerm = 0;
+              $keyTypeTerm $keyFieldTerm = ${rowFieldReadAccess(i, keys, keyType)};
+              $keyHashTerm = $keyElementHashBody;
+            }
+
+            boolean $valueIsNull = $values.isNullAt($i);
+            if(!$valueIsNull) {
+              $valueHashTerm = 0;
```
Comment (wuchong) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/HashCodeGenerator.scala:208]: Should the `$keyHashTerm = 0;` and `$valueHashTerm = 0;` be executed before if condition? Otherwise, if current value is null, the `valueHashTerm` would be the hash of the last entry. Could you add a test for a map with multiple null values?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/HashCodeGenerator.scala [hunk 6]
```diff
@@ -75,6 +75,157 @@ object HashCodeGenerator {
...
+
+        ${ctx.reuseMemberCode()}
+
+        public $className(Object[] references) throws Exception {
+          ${ctx.reuseInitCode()}
+        }
+
+        @Override
+        public int hashCode($ARRAY_DATA $inputTerm) {
+          int $hashIntTerm = 0;
```
Comment (wuchong) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/HashCodeGenerator.scala:123]: Should the `int $hashIntTerm = 0;` start from `1`?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/HashCodeGenerator.scala [hunk 7]
```diff
@@ -75,6 +75,157 @@ object HashCodeGenerator {
...
+          // This is inspired by hive & presto
+          for (int $i = 0; $i < $inputTerm.size(); $i++) {
+            boolean $isNull = $inputTerm.isNullAt($i);
+            if (!$isNull) {
+              $typeTerm $fieldTerm = ${rowFieldReadAccess(i, inputTerm, elementType)};
+              $hashIntTerm = 31 * $hashIntTerm + $elementHashBody;
+            }
```
Comment (wuchong) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/HashCodeGenerator.scala:130]: Should we consider null elements, e.g. use hash 0 for null elements instead of skipping them?

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/generated/HashFunction.java [hunk 8]
```diff
@@ -18,12 +18,23 @@
...
 import org.apache.flink.table.data.RowData;
 
 /**
- * Interface for code generated hash code of {@link RowData}, which will select some fields to hash.
+ * Interface for code generated hash code of {@link RowData} which will select some fields to hash
+ * or {@link ArrayData} or {@link MapData}.
+ *
+ * <p>Due to Janino's support for generic type is not very friendly, so here can't introduce generic
+ * type for {@link HashFunction}, please see https://github.com/janino-compiler/janino/issues/109
+ * for details.
```
Comment (wuchong) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/generated/HashFunction.java:31]: You can define generic types in the interface. The Janino limitation requires generating the method parameters using `Object` types instead of generic types. However, that's not a problem. You can manually cast the input parameter to `RowData`/`ArrayData`/`MapData` type in the first line of the method. See `org.apache.flink.table.planner.codegen.FunctionCodeGenerator#generateFunction`

I don't like overloading many methods, but just one method is valid.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/codegen/HashCodeGeneratorTest.scala [hunk 9]
```diff
@@ -41,16 +45,164 @@ class HashCodeGeneratorTest {
...
+    // test primitive type
+    val hashFunc1 = HashCodeGenerator
+      .generateArrayHash(
+        new CodeGeneratorContext(new Configuration(), classLoader),
+        new IntType(),
+        "name")
+      .newInstance(classLoader)
+
+    val array1 = new GenericArrayData(Array(1, 5, 7))
+    assertEquals(1123, hashFunc1.hashCode(array1))
```
Comment (wuchong) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/codegen/HashCodeGeneratorTest.scala:107]: Asserting the hash int is hard to tell right or wrong. Would be better to add assertions, e.g. 

```
val array1 = new GenericArrayData(Array(1, 5, 7))
    val array2 = new GenericArrayData(Array(1, 5, 7))
    val array3 = new GenericArrayData(Array[AnyRef](1, null, 5, null, 7))
    assertEquals(hashFunc1.hashCode(array1), hashFunc1.hashCode(array2))
    assertNotEquals(hashFunc1.hashCode(array1), hashFunc1.hashCode(array3))
```

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/codegen/HashCodeGeneratorTest.scala [hunk 10]
```diff
@@ -41,16 +45,164 @@ class HashCodeGeneratorTest {
...
+    val hashFunc1 = HashCodeGenerator
+      .generateMapHash(
+        new CodeGeneratorContext(new Configuration(), classLoader),
+        new IntType(),
+        new VarCharType(),
+        "name")
+      .newInstance(classLoader)
+
+    val map1 = new GenericMapData(
+      Map(1 -> StringData.fromString("ron"), 5 -> StringData.fromString("danny"), 10 -> null))
```
Comment (wuchong) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/codegen/HashCodeGeneratorTest.scala:161]: Please test hashcode of 2 maps are equal, which have entries in a different order and contain null values.

**Review by wuchong — COMMENTED** [review]
Please add IT cases for `SELECT DISTINCT <array_type_column>` and `SELECT COUNT(DISTINCT <array_type_column)` as well.

**Comment by lsyldliu** [comment]
@wuchong Thanks for review, I've addressed your comments.

**Review by wuchong — APPROVED** [review]
LGTM.

**Comment by wuchong** [comment]
The failed case is tracked by FLINK-26721
